client/extensions/woocommerce/app/settings/shipping/shipping-zone/shipping-zone-method-dialog.js

```diff
@@ -0,0 +1,178 @@
+/**
+ * External dependencies
+ */
+import React, { PropTypes } from 'react';
+import { bindActionCreators } from 'redux';
+import { connect } from 'react-redux';
+import { localize } from 'i18n-calypso';
+import Gridicon from 'gridicons';
+
+/**
+ * Internal dependencies
+ */
+import Dialog from 'components/dialog';
+import FlatRate from './shipping-methods/flat-rate';
+import FormFieldSet from 'components/forms/form-fieldset';
+import FormLabel from 'components/forms/form-label';
+import FormSelect from 'components/forms/form-select';
+import FormTextInput from 'components/forms/form-text-input';
+import FormToggle from 'components/forms/form-toggle/compact';
+import FreeShipping from './shipping-methods/free-shipping';
+import LocalPickup from './shipping-methods/local-pickup';
+import {
+	getShippingMethodNameMap,
+} from 'woocommerce/state/sites/shipping-methods/selectors';
+import {
+	changeShippingZoneMethodTitle,
+	changeShippingZoneMethodType,
+	cancelShippingZoneMethod,
+	closeShippingZoneMethod,
+	removeMethodFromShippingZone,
+	toggleOpenedShippingZoneMethodEnabled,
+} from 'woocommerce/state/ui/shipping/zones/methods/actions';
+import {
+	getMethodTypeChangeOptions,
+	getCurrentlyOpenShippingZoneMethod,
+	isCurrentlyOpenShippingZoneMethodNew,
+} from 'woocommerce/state/ui/shipping/zones/methods/selectors';
+
+const ShippingZoneDialog = ( {
+		siteId,
+		method,
+		methodNamesMap,
+		methodTypeOptions,
+		translate,
+		isVisible,
+		isNew,
+		onChange,
+		actions
+	} ) => {
+	if ( ! isVisible ) {
+		return null;
+	}
+
+	const { enabled, methodType, title } = method;
+	const onCancel = () => {
+		if ( isNew ) {
+			actions.removeMethodFromShippingZone( siteId, method.id );
```

Comment:
This logic is wrong I think. You should `cancelShippingZoneMethod` regardless. If the method hasn't been comitted yet to the zone changes, it will be discarded in the reducer, which is what we want. With this line, if you add a method, close the dialog, then you open the dialog again and you cancel, it will be removed.